### PR TITLE
Miscellaneous updates

### PR DESCRIPTION
**Pull request by Sssssaltyfish**
1. Fixed OpenSBI used by riscv64 qemu by providing a prebuilt binary.
2. Fixed multiple problems relating to the building procedure.
3. Added multiple features that enhance developer experience.

### CODE REVIEW

**Comment by coolyjg** [comment]
@Harris-pku This PR provides a pre-build binary for openSBI. Try this.

File: scripts/make/build_c.mk [hunk 1]
```diff
@@ -65,12 +67,16 @@ app-objs := main.o
...
+
 $(APP)/%.o: $(APP)/%.c $(ulib_hdr)
 	$(call run_cmd,$(CC),$(CFLAGS) $(APP_CFLAGS) -c -o $@ $<)
 
+$(rust_lib): _cargo_build
+
 $(OUT_ELF): $(c_lib) $(rust_lib) $(libgcc) $(app-objs)
 	@printf "    $(CYAN_C)Linking$(END_C) $(OUT_ELF)\n"
-	$(call run_cmd,$(LD),$(LDFLAGS) $^ -o $@)
+	$(call run_cmd,$(LD),$(LDFLAGS) $(c_lib) $(rust_lib) $(libgcc) $(app-objs) -o $@)
```
Comment (coolyjg) [inline scripts/make/build_c.mk:79]: Is this different from `^`?

File: scripts/make/build_musl.mk [hunk 2]
```diff
@@ -60,9 +60,11 @@ app-objs := $(addprefix $(APP)/,$(app-objs))
 $(APP)/%.o: $(APP)/%.c build_musl
 	$(call run_cmd,$(CC),$(CFLAGS) $(APP_CFLAGS) -c -o $@ $<)
 
+$(rust_lib): _cargo_build
+
 $(OUT_ELF): $(c_lib) $(rust_lib) $(libgcc) $(app-objs)
 	@printf "    $(CYAN_C)Linking$(END_C) $(OUT_ELF)\n"
-	$(call run_cmd,$(LD),$(LDFLAGS) $^ -o $@)
+	$(call run_cmd,$(LD),$(LDFLAGS) $(c_lib) $(rust_lib) $(libgcc) $(app-objs) -o $@)
```
Comment (coolyjg) [inline scripts/make/build_musl.mk:67]: Also here, is it different from `^`?

File: crates/axio/src/buffered/bufwriter.rs [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+///     stream.write(&[i+1]).unwrap();
+/// }
+/// stream.flush().unwrap();
+/// ```
+///
+/// By wrapping the stream with a `BufWriter<W>`, these ten writes are all grouped
+/// together by the buffer and will all be written out in one system call when
+/// the `stream` is flushed.
+///
+/// [`TcpStream::write`]: super::super::super::net::TcpStream::write
```
Comment (coolyjg) [inline crates/axio/src/buffered/bufwriter.rs:67]: Line 67-68 cause doc check fail.

File: modules/ruxfs/src/api/dir.rs [hunk 4]
```diff
@@ -150,10 +150,7 @@ impl DirBuilder {
         }
     }
 
-    fn create_dir_all(&self, _path: &str) -> Result<()> {
-        axerrno::ax_err!(
-            Unsupported,
-            "Recursive directory creation is not supported yet"
-        )
+    pub fn create_dir_all(&self, path: &str) -> Result<()> {
```
Comment (coolyjg) [inline modules/ruxfs/src/api/dir.rs:153]: Add comment doc for this function.

**Review by coolyjg — COMMENTED** [review]
This PR has done a great job, especially:
- Supporting creating files and directories recursively.
- Adjust some build bugs in MAKEFILE.
- More debug messages for scheduler and error handling.
- Fixed Riscv OpenSBI binary.
- Formalize some API in ulib/axstd.

However, for the BufferWriter and BufferReader part, I need more time to review related code. Copying from Rust std is acceptable, but comments are not proper.

I think it would be better to move those code into another PR. @ken4647

File: crates/linked_list/src/linked_list.rs [hunk 5]
```diff
@@ -214,6 +214,11 @@ impl<G: AdapterWrapped> List<G> {
     pub fn cursor_front(&self) -> Cursor<'_, G> {
         self.list.cursor_front()
     }
+
+    /// Returns an iterator of its elements.
+    pub fn iter(&self) -> Iterator<'_, G> {
```
[the remark this answers is not in the document]
Comment (Sssssaltyfish) [reply crates/linked_list/src/linked_list.rs:219]: Solved

File: scripts/make/build_c.mk (hunk 1 above)
Comment (Sssssaltyfish) [reply scripts/make/build_c.mk:79]: `$(rust_lib)` depends on `_cargo_build`, which is a .PHONY target. `$^` does not take this into consideration, and will cause `_cargo_build` to appear in the final substitution, leading to build failure.

File: scripts/make/build_musl.mk (hunk 2 above)
Comment (Sssssaltyfish) [reply scripts/make/build_musl.mk:67]: See comments in `build_c.mk`.

File: modules/ruxfs/src/api/dir.rs (hunk 4 above)
Comment (Sssssaltyfish) [reply modules/ruxfs/src/api/dir.rs:153]: Solved.

File: crates/axerrno/src/lib.rs [hunk 6]
```diff
@@ -138,6 +138,11 @@ macro_rules! ax_err_type {
         $crate::__priv::warn!("[AxError::{:?}] {}", $err, $msg);
         $err
     }};
+    ($err: ident, $fmt: literal, $($arg: tt)+) => {{
```
[the remark this answers is not in the document]
Comment (Sssssaltyfish) [reply crates/axerrno/src/lib.rs:141]: Solved.

File: crates/axio/src/buffered/bufwriter.rs (hunk 3 above)
Comment (Sssssaltyfish) [reply crates/axio/src/buffered/bufwriter.rs:67]: Changed reference to `std`.

**Comment by Sssssaltyfish** [comment]
> This PR has done a great job, especially:
> 
> * Support creating files and directories recursively.
> * Fix some build bugs in MAKEFILE.
> * Add more debug messages for scheduler and error handling.
> * Fixed Riscv OpenSBI binary.
> * Formalize some API in ulib/axstd.
> 
> However, for the BufferWriter and BufferReader part, I need more time to review related code. Copying from Rust std is acceptable, but comments are not proper.
> 
> I think it would be better to move those code into another PR. @ken4647
> 
> By the way, please use `make clippy`, `make doc_check_missing`, `make test` before handling the code.

Should I remove the `BufReader`/`BufWriter` from this PR for now?

**Comment by coolyjg** [comment]
> > This PR has done a great job, especially:
> > 
> > * Support creating files and directories recursively.
> > * Fix some build bugs in MAKEFILE.
> > * Add more debug messages for scheduler and error handling.
> > * Fixed Riscv OpenSBI binary.
> > * Formalize some API in ulib/axstd.
> > 
> > However, for the BufferWriter and BufferReader part, I need more time to review related code. Copying from Rust std is acceptable, but comments are not proper.
> > I think it would be better to move those code into another PR. @ken4647
> > By the way, please use `make clippy`, `make doc_check_missing`, `make test` before handling the code.
> 
> Should I remove the `BufReader`/`BufWriter` from this PR for now?

Yes, and put them into another PR. And for this pr, in order to make every commit clean and concise, please rebase this PR into one commit. Thanks!

**Comment by coolyjg** [comment]
By the way, please fix CI as well.

**Review by ken4647 — APPROVED** [review]

**Review by ken4647 — CHANGES_REQUESTED** [review]
The message of commit should be more meaningful. Consider to use some useful hint insteal of "parent".

File: modules/ruxhal/src/arch/x86_64/mod.rs [hunk 7]
```diff
@@ -19,6 +19,8 @@ use core::arch::asm;
 use memory_addr::{PhysAddr, VirtAddr};
 use x86::{controlregs, msr, tlb};
 use x86_64::instructions::interrupts;
+
+#[allow(unused_imports)]
```
Comment (ken4647) [inline modules/ruxhal/src/arch/x86_64/mod.rs:23]: I think use `#[cfg(feature = "musl")]` would be more clearness than `#[allow(unused_imports)]`.

Comment (Sssssaltyfish) [reply modules/ruxhal/src/arch/x86_64/mod.rs]: Adjusted.

**Review by ken4647 — APPROVED** [review]

**Comment by Sssssaltyfish** [comment]
> The message of commit should be more meaningful. Consider to use some useful hint insteal of "parent".

Now it should be better.

**Comment by coolyjg** [comment]
This could be merged after CI is passed (also check if all warnings are solved).

**Pull request merged**